File: Doubly_linked_list/DoublyLinkedListTraversal.java
// File Name: DoublyLinkedListTraversal.java
// Program to traverse a doubly linked list left to right and right to left

class Node {
    int data;
    Node prev;
    Node next;

    Node(int data) {
        this.data = data;
        this.prev = null;
        this.next = null;
    }
}

public class DoublyLinkedListTraversal {
    static Node head = null;

    // Traverse left to right
    static void traverseForward() {
        if (head == null) {
            System.out.println("List is empty");
            return;
        }
        Node temp = head;
        System.out.print("Forward traversal: ");
        while (temp != null) {
            System.out.print(temp.data + " <-> ");
            temp = temp.next;
        }
        System.out.println("NULL");
    }

    // Traverse right to left
    static void traverseBackward() {
        if (head == null) {
            System.out.println("List is empty");
            return;
        }
        Node temp = head;
        while (temp.next != null)
            temp = temp.next;

        System.out.print("Backward traversal: ");
        while (temp != null) {
            System.out.print(temp.data + " <-> ");
            temp = temp.prev;
        }
        System.out.println("NULL");
    }

    public static void main(String[] args) {
        traverseForward();
        traverseBackward();
    }
}

File: Doubly_linked_list/doubly_linked_list_traversal.py
# File Name: doubly_linked_list_traversal.py
# Program to traverse a doubly linked list left to right and right to left

class Node:
    def __init__(self, data):
        self.data = data
        self.prev = None
        self.next = None

head = None

# Traverse left to right
def traverse_forward():
    global head
    if head is None:
        print("List is empty")
        return
    temp = head
    print("Forward traversal:", end=" ")
    while temp:
        print(temp.data, end=" <-> ")
        temp = temp.next
    print("NULL")

# Traverse right to left
def traverse_backward():
    global head
    if head is None:
        print("List is empty")
        return
    temp = head
    while temp.next:
        temp = temp.next
    print("Backward traversal:", end=" ")
    while temp:
        print(temp.data, end=" <-> ")
        temp = temp.prev
    print("NULL")

if __name__ == "__main__":
    traverse_forward()
    traverse_backward()
